### PR TITLE
Add Facebook Handover Protocol 

### PR DESCRIPTION
**Pull request by ouadie-lahdioui**
Hello

This PR add FB [Messenger handover protocol](https://developers.facebook.com/docs/messenger-platform/handover-protocol) that enables two or more applications to collaborate on the Messenger platform for a page.

Enjoy ...

### CODE REVIEW

**Comment by jonchurch** [comment]
Just submitted a PR to your fork @ouadie-lahdioui to bring this up to v0.6 compatibility

Awesome work though btw! FB keeps pumping out new Webhook events like crazy these days haha

File: lib/Facebook.js [hunk 1]
```diff
@@ -763,6 +795,87 @@ function Facebookbot(configuration) {
...
+                    if (err) {
+                        facebook_botkit.log('Could not pass thread control');
+                    } else {
+                        if (body) {
+                            if (body.error) {
+                                facebook_botkit.log('ERROR in pass thread control API call: ', body.error.message);
+								cb && cb(body)
+                            } else {
+                                facebook_botkit.debug('Successfully pass thread control', body);
+                                cb && cb(null, body);
```
Comment (t-lopes) [inline lib/Facebook.js]: Looks like some code is missing here

**Comment by deubaka** [comment]
Any updates on this? Interested as to not duplicate effort. :)

File: lib/Facebook.js (hunk 1 above)
Comment (ouadie-lahdioui) [reply lib/Facebook.js]: Fixed

File: lib/Facebook.js [hunk 2]
```diff
@@ -458,6 +465,26 @@ function Facebookbot(configuration) {
...
+	facebook_botkit.middleware.categorize.use(function threadControl(bot, message, next) {
+
+		if (message.app_roles) {
+			message.type = 'facebook_app_roles';
+		} 
+		if (message.standby) {
+			message.type = 'standby'
+		}
+		if (message.pass_thread_control) {
+			message.type = 'facebook_receive_thread_control'
```
[the remark this answers is not in the document]
Comment (ouadie-lahdioui) [reply lib/Facebook.js]: pass_thread_control callback will occur when thread ownership for a user has been passed to your application, So, in botkit y'll be able to catche that like that :
````
controller.hears('facebook_receive_thread_control', 'message_received', function(bot, message) { 
 /* your bot has received the thread ownership */
});
````

However, take_thread_control will occur when thread ownership for a user has been taken away from your application, to catche that : 
````
controller.hears('facebook_lose_thread_control', 'message_received', function(bot, message) { 
 /* thread ownership for a user has been taken away */
});
````

@t-lopes WDYT ?

**Pull request merged**